### PR TITLE
Implement no_grad graph suppression

### PR DESCRIPTION
**Pull request by jdblackstar**
## Summary
- Make `Tensor.no_grad()` suppress autograd graph creation for tensor operations.
- Return operation results with `requires_grad=False` while no-grad mode is active.
- Preserve normal gradient tracking after leaving the context, including nested contexts and decorator use.

## Tests
- `uv run pytest tests/test_tensor.py tests/test_ops.py -q` (blocked: uv cache permission for `/Users/josh/.cache/uv`)
- `/Users/josh/code/punytorch/.venv/bin/python -m pytest tests/test_tensor.py tests/test_ops.py -q` (11 passed in 0.05s)

---

> [!NOTE]
> **Medium Risk**
> Touches core autograd wiring for many tensor paths; incorrect gating could silently break training or leak graphs, though behavior is covered by new tests.
> 
> **Overview**
> **`Tensor.no_grad()`** now actually stops building the autograd graph: a class-level **`_compute_grad`** flag is toggled in the context manager (with **nested restore** of the previous value), and a new **`_should_track_grad`** helper gates whether ops set **`requires_grad`** and attach **`Function`** contexts.
> 
> Operations across indexing, transpose, reductions, binary ops, activations, reshape, clone, **`long`**, and **`cross_entropy`** were switched from checking only per-tensor **`requires_grad`** to honoring **`_compute_grad`**. The **`no_grad`** decorator was fixed to use **`@wraps`**, enter via **`Tensor.no_grad()`**, and return **`self`** from **`__enter__`**.
> 
> New tests cover context and decorator behavior, nested contexts, and that backprop still works after leaving **`no_grad`**.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 1fe61bb8c59655feaf514d6c403917822a515734. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by jdblackstar** [comment]
closing, superseded by https://github.com/jdblackstar/punytorch/pull/14
